x-pack/plugins/apm/public/components/shared/FilterBar/ServiceEnvFilter.tsx

```diff
@@ -0,0 +1,114 @@
+import { useFetcher } from '../../../hooks/useFetcher';
+import { loadServiceEnvironments } from '../../../services/rest/apm/services';
+import { useLocation } from '../../../hooks/useLocation';
+import { useUrlParams } from '../../../hooks/useUrlParams';
+import { history } from '../../../utils/history';
+import { fromQuery, toQuery } from '../Links/url_helpers';
+
+const anyEnvironmentValue = 'Any';
+const nullEnvironmentValue = 'Null';
+
+function updateEnvironmentUrl(
+  location: ReturnType<typeof useLocation>,
+  environment?: string
+) {
+  const nextEnvironmentQueryParam =
+    environment !== anyEnvironmentValue ? environment : undefined;
+  history.push({
+    ...location,
+    search: fromQuery({
+      ...toQuery(location.search),
+      environment: nextEnvironmentQueryParam
+    })
+  });
+}
+
+const OptionInputContainer = styled.div`
+  min-width: 192px;
+`;
+
+const EnvironmentLabel = styled.span`
+  color: #adb2b9;
+  margin-right: 6px;
+`;
+
+const OptionInput: React.FC = ({ children }) => {
+  return (
+    <OptionInputContainer>
+      <EnvironmentLabel>
+        {i18n.translate('xpack.apm.filter.environment.label', {
+          defaultMessage: 'environment'
+        })}
+        :
+      </EnvironmentLabel>{' '}
+      {children}
+    </OptionInputContainer>
+  );
+};
+
+export const ServiceEnvFilter: React.FC = () => {
+  const location = useLocation();
+  const { urlParams, uiFilters } = useUrlParams();
+  const { start, end } = urlParams;
+  const {
+    global: { environment }
+  } = uiFilters;
+  const { data = [], status = 'loading' } = useFetcher(
+    () => {
+      if (start && end) {
+        return loadServiceEnvironments({ start, end });
+      }
+    },
+    [start, end]
+  );
+
+  const availableEnvironments = [
+    {
+      value: anyEnvironmentValue,
+      label: i18n.translate('xpack.apm.filter.environment.anyLabel', {
+        defaultMessage: 'Any'
+      })
+    },
+    ...data.map(environment => ({ value: environment, label: environment })),
+    {
+      value: nullEnvironmentValue,
+      label: i18n.translate('xpack.apm.filter.environment.notSetLabel', {
+        defaultMessage: 'Not set'
+      })
+    }
+  ];
+
+  if (
+    environment &&
+    environment !== anyEnvironmentValue &&
+    environment !== nullEnvironmentValue &&
+    status === 'loading'
+  ) {
+    availableEnvironments.push({ value: environment, label: environment });
+  }
+
+  const [valueOfSelected, setValueOfSelected] = useState(
+    environment || availableEnvironments[0].value
+  );
+
+  return (
+    <EuiSuperSelect
+      options={availableEnvironments.map(({ value, label }) => ({
+        value,
+        inputDisplay: <OptionInput>{label}</OptionInput>,
+        dropdownDisplay: <>{label}</>
+      }))}
+      valueOfSelected={valueOfSelected}
+      onChange={(value: string) => {
+        setValueOfSelected(value);
+        updateEnvironmentUrl(location, value);
+      }}
+      isLoading={status === 'loading'}
```

Comment:
Use constants: `FETCH_STATUS.LOADING`

Reply:
Actually, I might take that back. If typescript validates that `status` must be a valid `FETCH_STATUS` then I guess using strings is okay.